scipy/spatial/src/distance_metrics.h

```diff
@@ -0,0 +1,128 @@
+#pragma once
+
+#include <cmath>
+#include "views.h"
+
+struct MinkowskiDistance {
+    double p_;
+
+    template <typename T>
+    void operator()(StridedView2D<T> out, StridedView2D<const T> x, StridedView2D<const T> y) const {
+        T p = p_;
+        for (intptr_t i = 0; i < x.shape[0]; ++i) {
+            T p_dist = 0;
+            for (intptr_t j = 0; j < x.shape[1]; ++j) {
+                auto diff = std::abs(x(i, j) - y(i, j));
+                p_dist += std::pow(diff, p);
+            }
+            out(i, 0) = std::pow(p_dist, 1 / p);
+        }
+    }
+
+    template <typename T>
+    void operator()(StridedView2D<T> out, StridedView2D<const T> x, StridedView2D<const T> y, StridedView2D<const T> w) const {
+        T p = p_;
+        for (intptr_t i = 0; i < x.shape[0]; ++i) {
+            T p_dist = 0;
+            for (intptr_t j = 0; j < x.shape[1]; ++j) {
+                auto diff = std::abs(x(i, j) - y(i, j));
+                p_dist += w(i, j) * std::pow(diff, p);
+            }
+            out(i, 0) = std::pow(p_dist, 1 / p);
+        }
+    }
+};
+
+struct EuclideanDistance {
+    template <typename T>
+    void operator()(StridedView2D<T> out, StridedView2D<const T> x, StridedView2D<const T> y) const {
+        for (intptr_t i = 0; i < x.shape[0]; ++i) {
+            T sqdist = 0;
+            for (intptr_t j = 0; j < x.shape[1]; ++j) {
+                auto diff = std::abs(x(i, j) - y(i, j));
+                sqdist += diff * diff;
+            }
+            out(i, 0) = std::sqrt(sqdist);
+        }
+    }
+
+    template <typename T>
+    void operator()(StridedView2D<T> out, StridedView2D<const T> x, StridedView2D<const T> y, StridedView2D<const T> w) const {
+        for (intptr_t i = 0; i < x.shape[0]; ++i) {
+            T sqdist = 0;
+            for (intptr_t j = 0; j < x.shape[1]; ++j) {
+                auto diff = std::abs(x(i, j) - y(i, j));
+                sqdist += w(i, j) * (diff * diff);
+            }
+            out(i, 0) = std::sqrt(sqdist);
+        }
+    }
+};
+
+struct ChebyshevDistance {
+    template <typename T>
+    void operator()(StridedView2D<T> out, StridedView2D<const T> x, StridedView2D<const T> y) const {
+        for (intptr_t i = 0; i < x.shape[0]; ++i) {
+            T dist = 0;
+            for (intptr_t j = 0; j < x.shape[1]; ++j) {
+                auto diff = std::abs(x(i, j) - y(i, j));
+                if (diff > dist) {
+                    dist = diff;
+                }
+            }
+            out(i, 0) = dist;
+        }
+    }
+
+    template <typename T>
+    void operator()(StridedView2D<T> out, StridedView2D<const T> x, StridedView2D<const T> y, StridedView2D<const T> w) const {
+        for (intptr_t i = 0; i < x.shape[0]; ++i) {
+            T dist = 0;
+            for (intptr_t j = 0; j < x.shape[1]; ++j) {
+                auto diff = std::abs(x(i, j) - y(i, j));
+                if (w(i, j) > 0 && diff > dist) {
```

Comment:
Is the docstring for 1D `chebyshev` incorrect in indicating that weights are:

> Unused, as ‘max’ is a weightless operation. Here for API consistency.

It looks like something does happen with weights even in the "1D" Python source; perhaps that's why it does here as well.

I don't know if this what one would intuitively expect a "weight" to do--probably not--I suspect we might agree on that and this is to replicate the old behavior? The Euclidean weight above makes a lot more sense.

Reply:
This is consistent with the existing metric functions, so I don't think this is the right place to discuss whether it's correct.

That said, I interpret integer weights as "what would the distance be if there were `w` dimensions all with this value". With that interpretation, it's clear that zero weight entries should have no effect on the result. If anything, I think the docs should be changed.

Reply:
hmm, it wasn't clear to me from the docstring that `w` was integer only since it is categorized as `array_like` there. Either way, docstring is wrong because clearly the `w` can be used, but this is indeed not the place to discuss.

Reply:
Perhaps that was confusing, I didn't mean that `w` is an integer here it's just that the model of having `w` equal dimensions only makes sense for integer values.